### ISSUE TITLE
Default value breaks types

### ISSUE THREAD

**Issue by jansedlon** [edmundhung/conform#478]
### Describe the bug and the expected behavior

When using for example zod discriminated union for form schema, providing default value breaks the types completely.

### Conform version

v1.0.2

### Steps to Reproduce the Bug or Issue

https://stackblitz.com/edit/github-qiy4lh-vn8142?file=app%2Froutes%2Ftype-inference.tsx

**Comment by jansedlon** [comment]
From what I understand, the `defaultValue` must exactly match the schema type which is a bit cumbersome and breaks other things.

A workaround I found is to fallback all string values to `?? ""` even if they are optional, but that's cumbersome as well 😅

**Comment by edmundhung** [comment]
It looks like Typescript is trying to infer the types from both `defaultValue` and `onValidate`. I think this will be best solved by the [NoInfer](https://devblogs.microsoft.com/typescript/announcing-typescript-5-4-beta/#the-noinfer-utility-type) which is still on beta 😅. Let me check if there is another approach that can work with earlier version.

A temporary solution is to set the type manually to stop typescript from inferring it.
```tsx
const [form, fields] = useForm<z.input<typeof schema>>({
   // ...
});
```

**Comment by jansedlon** [comment]
Yeah I've already tried to apply NoInfer directly in the conform source code but to no avail 😂

**Comment by edmundhung** [comment]
Typescript seems to infer the type correctly if you spread over the `defaultValues` instead of passing it to the `useForm` hook directly:

```tsx
const [form, fields] = useForm({
    lastResult,
    onValidate({ formData }) {
      return parseWithZod(formData, { schema: LinkPageFormSchema });
    },
    defaultValue: {
      ...defaultValues
    },
});
```

**Comment by aaronadamsCA** [comment]
@edmundhung, I've done my best to create an isolated reproduction of the problems here: [TypeScript Playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBDAnmApnA3gKDnAIigMwEMBXAGxgDUiySUAabOAMWBTIBMBZFGIjon0Y5m0EDz4ChmAL5wCUCCDgAiAAIBjCADsCYgLQwIAeigoiGmCoDcmUJFgY4YIlADOKAOrAYACwBaEBxwcgpKqpo6elAghiYAXkE2duDQ8Ohw8SHyisoqiRzJmMYAVNglcACS4GQoICjafDDAOm7lxpiYBCTali3acCQeojEAPEwAyhq+dURwKAAeMA0cbnAASihaUByjbjBQwNoA5vSD2gDW2hAA7toAfMIsYtS0aAC8cFMzIERPIyAAKJQRRQOCffaHE4AbQAuox7gAKJgQMDNVoALmeMQA8mj+m5Rt9ZmcAa86I84MZjHBPL5EHAOBAUGs3DcfNMjscEDMPM5XER6st3PJgAs4DQQBB9vMAG4oKCIPxcgD8AEIVZgAJRY6FMAESfiCIhE6Yk7FAkHQR5MTYwEhQbQAFWQKFGBt4Rr4pp+fwtwNB92hKmOvFY7FWvBUsJtsIwTGpcDUMDc+kWqEsaatYP0+jg13g9lq9UaKA4TDM9sdcDhthknSOwuIGjQALx6O0hOJv1JLxodHBXzNv3u8ZwHEIpAo5JQKqx+GI5Co-bd3aI91sOB0r2AUlnWMRWlLSyxGWiIFwxqxAMvfBCWvBo4mJAARiBgG43P13X23hvZJ0SCoF8r7vp+35kiuo7vEwAA+Tj7IIQxYiobgkBoLafjYcCyiu16-gO9Y4PBGSIfabgoQqoIqHA8E9BOBBHGWIS2MUZRwBUTosim7SdIiD7vKOWA4EBaBcIga6DvEAB0RwEAqoyiRABBwG4w7rpuCCunA4mQW8UnSRAJAwGAxmKa6ymqep-5MEeMpqb6UlMDg0kcB+GiHO+2iCGWACq2j9IiKhuWpnlHII0AqGceo4LFmSGS+ABWWwwIiwlxbFIUeaA4VGFAWIyWQPgKjQQU6CgKhak8GXyBAEAFdJkJcvx1WxTIVXObFMkQElKVpZ1cVZWF3l5Q1RXCqVKgwDcECVa1ODta1sJap10kHEQnbnoiiKuMcAmjmlcDScdu1nC+rhYgAjAATAAzPeWqsTgdnwLhRUcAuU7LvpnzpeO7nDRF+WqOVUWdXo9WqDMZBkLNrXncDN23U8chEGsL1PXAL1wEcb27p9S4zoOf1wOd8QodDsNgwtEro60MCY90vQdljMwaBcAIuqgiIAPr4TEhpSCa4lrvcOr5uQZBjrFlYOgM2iS5p9ZMEzfQ6KzWwc2wnBcygyJxXzLDa9wnpC3sBxcrR5wMUxHBnCL1lPOLCsw9LOCy9WLtkErnQ4ImAByEBWY5olMPxj5u1j9M1ueZxnnVIRxp8QwoAC-U1QxX0zlieMfZOhMrq17WaX7NLJqm6YpVmoJwLmtUxHA7J+PmQcObMcC+GjLdaagnVmuznOuoi56PZ1ibl2mCwZjA1fQLXeaMRGjc+L43dt78Hdd9cPcoH3bNaxGuvD3Vo8LZjiaSaHODh4JkfY9CsdOBDieDinAKjA7vpIiTW7aDue4HnjjEW8cx2oRxcO4LwK9AgcGPsA40cdg7t0Wp1Yue9NaD25iPEuGsB7GyPhDU+LFfZUhpJUXQVFmLrzmFfOAN8hJ92jg-MQiDn4yCToMYYYh04ZW3DQXcPlAH1wvMae84DXAeG8H4GBcCRFCAQupe8RciHPX3pgvW2D0F4MPkPQhStz40kkhtYI54cIrh3m0a++075MMfvHIO7DX5cLGJ-c0ukCIoG-gNRk+dpx4TMe9Amfi3jzTgHw96gi6FALkaA6xEDJHQKCLIkBiDqFKNQSo3BHMxAELEJk-uB8da6JPvokhiZyHyRBFQxRxjhEBIHKJSxdDrEk3vnY2qDiOFv24T-Hxi5gl0BzvwvO-Tvp0FCeEgRywhHnhAWI2+8SoHSKSbMhBCjHIoLimguKBT1GyPyWo-BxSIBEJkAYocjlammLeg010awm6r1xsMvpX16m7ysRHVptiWFPwTo45OziQAfwkupe2iA9IUh4XFceKZJ7T1njmPMmclxvMZMyNYBY4C-BgNMJBvxvHIoGSgLETzAm+LGQwbxkyAFROEXMsBCyJFLICCssQKT1nII6lsg5GCcnFJiDy7RRSsElKYGcspZCKFVOCGkq5YhUWNOXs3Ulu4Xkopue85pnzGGdngMwmIrC-ldMBVC2KMKK5TyrlROeddCWvXMUyFkLd4DYtxdQgl5Ls4420LnIJFKJl-2GZEw6qy7wMtHIsqRLLYGhr9KRRRmy2qCuyTEXJAqcG7KOSKk5pTxVAA)

As you can see, in most places schema type inference is working _exactly opposite_ of what's expected. For example, if you hover over the last `useForm` call, you can see the first two type arguments are reversed; the inferred `Schema` type comes from `defaultValue`, while the inferred `FormValue` type comes from `schema`. This matches the behaviour I've been seeing locally.

Thankfully, passing explicit type arguments always works as expected, which I've also demonstrated in the reproduction. So for now, this is what we'll do in our codebase.

What's truly maddening is that on line 18, if you switch `FormOptions<Schema, FormValue>` to `FormOptions<FormValue, Schema>`, 6 of the 7 error cases are resolved! I have no idea why this works, because it _definitely_ shouldn't—it is absolutely backwards. And yet, somehow, it makes up for almost every incorrect type inference. 🤷
